src/platform/input/game-pads.js

```diff
@@ -68,62 +152,757 @@ const MAPS = {
+     * @param {GamePad} gamepad - The gamepad that was just connected.
+     * @example
+     * var onPadConnected = function (pad) {
+     *     if (!pad.mapping) {
+     *         // Map the gamepad as the system could not find the proper map.
+     *     } else {
+     *         // Make the gamepad pulse.
+     *     }
+     * };
+     * app.keyboard.on("gamepadconnected", onPadConnected, this);
      */
-    update() {
-        // move current buttons status into previous array
-        for (let i = 0, l = this.current.length; i < l; i++) {
-            const buttons = this.current[i].pad.buttons;
-            const buttonsLen = buttons.length;
-            for (let j = 0; j < buttonsLen; j++) {
-                if (this.previous[i] === undefined) {
-                    this.previous[i] = [];
-                }
-                this.previous[i][j] = buttons[j].pressed;
+
+    /**
+     * Fired when a gamepad is disconnected.
+     *
+     * @event GamePads#gamepaddisconnected
+     * @param {GamePad} gamepad - The gamepad that was just disconnected.
+     * @example
+     * var onPadDisconnected = function (pad) {
+     *     // Pause the game.
+     * };
+     * app.keyboard.on("gamepaddisconnected", onPadDisconnected, this);
+     */
+
+    set deadZone(value) {
+        deadZone = value;
+    }
+
+    /**
+     * Threshold for axes to return values. Must be between 0 and 1.
+     *
+     * @type {number}
+     * @ignore
+     */
+    get deadZone() {
+        return deadZone;
+    }
+
+    /**
+     * The list of previous buttons states.
+     *
+     * @type {boolean[][]}
+     * @ignore
+     */
+    get previous() {
+        const current = this.current;
+
+        for (let i = 0, l = current.length; i < l; i++) {
+            const buttons = current[i]._buttons;
+
+            if (!this._previous[i]) {
+                this._previous[i] = [];
+            }
+
+            for (let j = 0, m = buttons.length; j < m; j++) {
+                const button = buttons[i];
+                this._previous[i][j] = button ? button._previouslyPressed : false;
             }
         }
 
-        // update current
-        this.poll(this.current);
+        this._previous.length = this.current.length;
+        return this._previous;
+    }
+
+    /**
+     * Callback function when a gamepad is connecting.
+     *
+     * @param {GamepadEvent} event - The event containing the connecting gamepad.
+     * @ignore
+     */
+    _ongamepadconnected(event) {
+        const pad = new GamePad(event.gamepad, this.getMap(event.gamepad));
+        const current = this.current;
+
+        let padIndex = current.findIndex(gp => gp.index === pad.index);
+        while (padIndex !== -1) {
+            current.splice(padIndex, 1);
+            padIndex = current.findIndex(gp => gp.index === pad.index);
+        }
+
+        current.push(pad);
+        this.fire(EVENT_GAMEPADCONNECTED, pad);
+    }
+
+    /**
+     * Callback function when a gamepad is connection.
```

Comment:
```suggestion
     * Callback function when a gamepad is disconnecting.
```

Like this maybe? Or "disconnected"?